frontend/dockerfile/shell/lex.go

```diff
@@ -472,3 +494,111 @@ func BuildEnvs(env []string) map[string]string {
 
 	return envs
 }
+
+// convertShellPatternToRegex converts a shell-like wildcard pattern
+// (? is a single char, * either the shortest or longest (greedy) string)
+// to an equivalent regular expression.
+//
+// Based on
+// https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_13
+// but without the bracket expressions (`[]`)
+func convertShellPatternToRegex(pattern string, greedy bool) (*regexp.Regexp, error) {
+	var s scanner.Scanner
+	s.Init(strings.NewReader(pattern))
+	var out strings.Builder
+	out.Grow(len(pattern) + 4)
+
+	// match only at the beginning of the string
+	out.WriteByte('^')
+
+	// default: non-greedy wildcards
+	starPattern := ".*?"
+	if greedy {
+		starPattern = ".*"
+	}
+
+	for tok := s.Next(); tok != scanner.EOF; tok = s.Next() {
+		switch tok {
+		case '*':
+			out.WriteString(starPattern)
+			continue
+		case '?':
+			out.WriteByte('.')
+			continue
+		case '\\':
+			// } as part of ${} needs to be escaped, but the escape isn't part
+			// of the pattern
+			if s.Peek() == '}' {
+				continue
+			}
+			out.WriteRune('\\')
+			tok = s.Next()
+			if tok != '*' && tok != '?' && tok != '\\' {
+				return nil, errors.Errorf("invalid escape '\\%c'", tok)
+			}
+		// regex characters that need to be escaped
+		// escaping closing is optional, but done for consistency
+		case '[', ']', '{', '}', '.', '+', '(', ')', '|', '^', '$':
+			out.WriteByte('\\')
+		}
+		out.WriteRune(tok)
+	}
+	return regexp.Compile(out.String())
+}
+
+func trimPrefix(word, value string, greedy bool) (string, error) {
+	re, err := convertShellPatternToRegex(word, greedy)
+	if err != nil {
+		return "", errors.Errorf("invalid pattern (%s) in substitution: %s", word, err)
+	}
+
+	if idx := re.FindStringIndex(value); idx != nil {
+		value = value[idx[1]:]
+	}
+	return value, nil
+}
+
+// reverse without avoid reversing escapes, i.e. a\*c -> c\*a
+func reversePattern(pattern string) string {
+	patternRunes := []rune(pattern)
+	out := make([]rune, len(patternRunes))
+	lastIdx := len(patternRunes) - 1
+	for i := 0; i <= lastIdx; {
+		tok := patternRunes[i]
+		outIdx := lastIdx - i
+		if tok == '\\' && i != lastIdx {
+			out[outIdx-1] = tok
+			// the pattern is taken from a ${var#pattern}, so the last
+			// character can't be an escape character
+			out[outIdx] = patternRunes[i+1]
+			i += 2
+		} else {
+			out[outIdx] = tok
+			i++
+		}
+	}
+	return string(out)
+}
```

Comment:
This function is probably the last thing that makes me feel uncomfortable, but the solution to fix it is a bit too complex that I don't feel like it adds anything to this solution.

The reason this is uncomfortable for me is because it adds parsing logic into the reversal of the pattern which could potentially make it so that logic is duplicated in two places. The only way I can think of to remedy this is to separate the parsing of the shell pattern from the generation of the regular expression. Mocking out names:

```go
type shellPattern struct {
    components []shellPatternComponent
    reversed bool
}

func (p *shellPattern) toRegex() *regexp.Regexp {
    // logic for generating the regular expression
    // since each component is in its own object, it's easier to reverse
    // the pattern by just iterating in a different direction.
}
```

I'm not suggesting this. I think if this works and it has tests then it's good enough and I don't know if there's a substantial benefit to doing it in another way. I just wanted to get my thoughts down in case I come across this section of code in the future and want to remind myself of the context.

Reply:
One might also go to the next logical step and generate an AST from the whole expression and then just reverse the tokens for the pattern.

The reversal code is a less than 20 lines (not including the 11+ lines of unit tests for it). It could be shorter (and definitely easier to understand), but unless it's used in a lot more places I find it hard to justify an additional class + parser code.

One feature that would necessitate your proposed change would be character classes in patterns, e.g. `H[aeu]llo`, but I think the potential use is far outweighed by the complexity.